private/bufpkg/bufcheck/runner_provider.go

```diff
@@ -15,146 +15,47 @@
 package bufcheck
 
 import (
-	"context"
-	"sync"
-
-	"github.com/bufbuild/buf/private/bufpkg/bufconfig"
-	"github.com/bufbuild/buf/private/bufpkg/bufparse"
 	"github.com/bufbuild/buf/private/bufpkg/bufplugin"
 	"github.com/bufbuild/buf/private/pkg/pluginrpcutil"
 	"github.com/bufbuild/buf/private/pkg/syserror"
 	"github.com/bufbuild/buf/private/pkg/wasm"
 	"pluginrpc.com/pluginrpc"
 )
 
-type runnerProvider struct {
-	wasmRuntime        wasm.Runtime
-	pluginKeyProvider  bufplugin.PluginKeyProvider
-	pluginDataProvider bufplugin.PluginDataProvider
+type localRunnerProvider struct {
+	wasmRuntime wasm.Runtime
 }
 
-func newRunnerProvider(
+func newLocalRunnerProvider(
 	wasmRuntime wasm.Runtime,
-	pluginKeyProvider bufplugin.PluginKeyProvider,
-	pluginDataProvider bufplugin.PluginDataProvider,
-) *runnerProvider {
-	return &runnerProvider{
-		wasmRuntime:        wasmRuntime,
-		pluginKeyProvider:  pluginKeyProvider,
-		pluginDataProvider: pluginDataProvider,
+) *localRunnerProvider {
+	return &localRunnerProvider{
+		wasmRuntime: wasmRuntime,
 	}
 }
 
-func (r *runnerProvider) NewRunner(pluginConfig bufconfig.PluginConfig) (pluginrpc.Runner, error) {
-	switch pluginConfig.Type() {
-	case bufconfig.PluginConfigTypeLocal:
+func (r *localRunnerProvider) NewRunner(plugin bufplugin.Plugin) (pluginrpc.Runner, error) {
+	switch isLocal, isWasm := plugin.IsLocal(), plugin.IsWasm(); {
+	case isLocal && !isWasm:
 		return pluginrpcutil.NewLocalRunner(
-			pluginConfig.Name(),
-			pluginConfig.Args()...,
+			plugin.Name(),
+			plugin.Args()...,
 		), nil
-	case bufconfig.PluginConfigTypeLocalWasm:
-		return pluginrpcutil.NewLocalWasmRunner(
+	case isLocal && isWasm:
+		return pluginrpcutil.NewWasmRunner(
 			r.wasmRuntime,
-			pluginConfig.Name(),
-			pluginConfig.Args()...,
+			plugin.Data,
+			plugin.Name(),
+			plugin.Args()...,
 		), nil
-	case bufconfig.PluginConfigTypeRemoteWasm:
-		return newRemoteWasmPluginRunner(
+	case !isLocal && isWasm:
+		return pluginrpcutil.NewWasmRunner(
 			r.wasmRuntime,
-			r.pluginKeyProvider,
-			r.pluginDataProvider,
-			pluginConfig.Ref(),
-			pluginConfig.Args(),
-		)
+			plugin.Data,
+			plugin.Name(),
+			plugin.Args()...,
+		), nil
```

Comment:
We don't really differentiate between local vs. remote Wasm plugins here anymore, we delegate that to `plugin.Data`.